### DISCUSSION TITLE
Merge two pages to one

### DISCUSSION THREAD

**Discussion by aashishvanand** [pymupdf/PyMuPDF#1130] [Looking for help]
Hi,
I have a Digitally signed pdf (PAdES) signed by user 1. There is a situation where I need to add some basic annotation and do an incremental save (so that I don't lose the 1st signature) and sign the same pdf with user 2 digital signature. 

I have the signed pdf and separate pdf with all the annotations at the correct place. Assuming both the pdf page size is the same is it possible to merge the two pages into one.

I am aware of the insert page in Fitz but I am looking for a merge page function rather than insert

`import fitz`
`doc1 = fitz.open("sample1_signed.pdf")`
`doc2 = fitz.open("sample2_annotation.pdf")`
`doc2.insertPDF(doc1)`
`doc2.save("document-Signed.pdf", incremental=True, encryption=fitz.PDF_ENCRYPT_KEEP)`

I am looking for something that is equivalent to PyPDF2 mergePage function [PyPDF2](https://pythonhosted.org/PyPDF2/PageObject.html#PyPDF2.pdf.PageObject.mergePage). Unfortunately PyPDF2 does not support incremental updates. 

`from PyPDF2 import PdfFileMerger, PdfFileReader, PdfFileWriter`
`file1 = PdfFileReader(open("sample1_signed.pdf", "rb"))`
`file2 = PdfFileReader(open("sample2_annotations.pdf", "rb"))`
`output = PdfFileWriter()`
`page = file1.getPage(0)`
`page.mergePage(file2.getPage(0))`
`output.addPage(page)`
`outputStream = open("document-output.pdf", "wb")`
`output.write(outputStream)`
`outputStream.close()`

**Comment by JorjMcKie** [comment]
There is method `page.show_pdf_page()`.
This displays a given page from another PDF (document object) on `page`. You choose which rectangle ("clip") of the source page should be considered and which rectangle part of the target page should be covered:

```python
page.show_pdf_page(rect, docsrc, pno=0, keep_proportion=True, overlay=True, oc=0, rotate=0, clip=None)
```

**Reply by aashishvanand** [reply]
Any simple way to do this If the original document has n pages. is there a built-in function to do it for all the pages? or should I run a loop for each page? 

`import fitz`
`doc1 = fitz.open("sample1_signed.pdf")`
`page = doc1.load_page(0)`
`doc2 = fitz.open("sample2_annotation.pdf")`
`page.show_pdf_page(page.MediaBox,doc2, pno=0, keep_proportion=True, overlay=True, oc=0, rotate=0, clip=None)`
`doc1.save("sample1_signed.pdf", incremental=True, encryption=fitz.PDF_ENCRYPT_KEEP)`

**Reply by JorjMcKie** [reply]
>  should I run a loop for each page?

yes, should be easy enough.
This is a multi-purpose function - see the docu.
Your case is just one out of half a dozen other possibilities.

BTW, why don't you use `page.rect` as the target rect? This is the visible page part. Apart from the fact that this property `page.mediabox` has been renamed to lower case ...

**Reply by aashishvanand** [reply]
Thanks, works great! But there seems to be an issue with the first signature. Adobe acrobat complains that document has been altered or corrupted since its signed. Even when using an incremental pdf writer. Anyways to fix this or stop adobe from complaining?

<img width="516" alt="Acrobat Incrmental Issue" src="https://user-images.githubusercontent.com/12725526/125014785-edd27200-e08b-11eb-86f7-e5176f310e69.png">

**Comment by JorjMcKie** [comment]
There is `Document.get_sigflags()`.  Return whether the document contains signature fields, an optional PDF property. But you may check if it does contain a usable value:
![grafik](https://user-images.githubusercontent.com/8290722/125080569-f5e1e000-e092-11eb-97c9-20182cc3a4d1.png)

Otherwise I am afraid I do not understand entirely:
Do you mean **_it always happens_** even when using incremental updates exclusively (with PyMuPDF)?
If that's the case, there might be an issue in my code: I am updating the second item in the PDF trailer's `/ID` key - irrespective of whether an **_incremental_** save is executed or not.
And that may be incorrect!

**Reply by JorjMcKie** [reply]
Correction of my previous post:
I reviewed my code and it turned out my memory was incorrect: the only case when I do something with the `/ID` key of the PDF trailer is, when this information does not exist at all.
Then I create the `/ID`. In all other cases, MuPDF exclusively does something with it.

As a conclusion: your observation cannot be caused by any MuPDF action. Maybe the PDF was - although signed - already corrupted?

**Reply by JorjMcKie** [reply]
Another update:
I have taken a signed document and added a page using incremental save.
Got the same result / message with Adobe as you did.
Then I did another experiment and just added an annotation, followed by incremental save - _**et voilà:**_ everything worked just fine - the doc **_stayed signed._**

So using incremental save does not mean you can do just anything you like to the PDF. Permanent changes like adding a new page or overlaying a page with permanent objects (images, text, xobjects like with `show_pdf_page`) will still invalidate the signature.
Annotations by definition are not permanent - they are like dust which you can wipe away.

So if I understood your intention well, you need some logic that reads the annotations from that other PDF and re-inserts them in the signed PDF ... etc.

**Reply by aashishvanand** [reply]
Thanks for the update! 
Regarding your first comment: 
Document.get_sigflags() I get a 3 as a response from my signed document

Regarding your second comment:
Nope, My document stays signed until any annotation or other elements are added. I have verified that that. 
I have attached a sample signed document.

Regarding your third comment:
I tried a sample Caret annotation But still, I get that signature is invalid 

<img width="413" alt="Screenshot 2021-07-11 at 10 15 14 AM" src="https://user-images.githubusercontent.com/12725526/125183090-2fe8e880-e231-11eb-89a9-5fbaab4c3721.png">

`import fitz`
`doc1 = fitz.open("signed.pdf")`
`page = doc1[0] `
`r = fitz.Rect(72, 72, 220, 100)`
`t1 = u"Hi"`
`red = (1, 0, 0)`
`annot = page.add_caret_annot(r.tl)`
`annot.parent.insert_text(annot.rect.br + (10, -5), "%s annotation" % annot.type[1], color=red)`
`doc1.save("signed.pdf", incremental=True, encryption=fitz.PDF_ENCRYPT_KEEP)`

I am definitely sure I am missing something here. Can you help me out? Not sure how you got 
> "Then I did another experiment and just added an annotation, followed by incremental save - et voilà: everything worked just fine - the doc stayed signed."

My understanding was incremental updates don't mess up with the signed parts and I should be able to add annotations provided they don't mess with the original signed document. I want to add basic annotations such as 
- square
- circle
- line
- polygon
- polyline
- ink
- text
- image
- checkbox

My initial PoC was rather than adding annotation one by one. I had a pdf with the same page size as the original one and all the annotations printed already. All that I had to do was to overlay the annotation pdf over the original pdf to get the final output

Attachments: 
Original Singed.pdf - No alterations made to the document
[Original Signed.pdf](https://github.com/pymupdf/PyMuPDF/files/6796365/Original.Signed.pdf)

Signed_Caret_Annoation.pdf - Added caret annotation but the signature is invalid
[Signed_Caret_Annoation.pdf](https://github.com/pymupdf/PyMuPDF/files/6796366/Signed_Caret_Annoation.pdf)

**Reply by aashishvanand** [reply]
Hi @JorjMcKie, is it possible for you to share the code used in your experiment?

> "Then I did another experiment and just added an annotation, followed by incremental save - et voilà: everything worked just fine - the doc stayed signed."
